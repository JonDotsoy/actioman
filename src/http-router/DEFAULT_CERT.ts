export const DEFAULT_CERT =
  "-----BEGIN CERTIFICATE-----\n" +
  "MIIDCTCCAfGgAwIBAgIUcx4/uMJz+em1t9pyg8YBn8mOtz8wDQYJKoZIhvcNAQEL\n" +
  "BQAwFDESMBAGA1UEAwwJbG9jYWxob3N0MB4XDTI1MDIwNzE1MDcyMloXDTI1MDMw\n" +
  "OTE1MDcyMlowFDESMBAGA1UEAwwJbG9jYWxob3N0MIIBIjANBgkqhkiG9w0BAQEF\n" +
  "AAOCAQ8AMIIBCgKCAQEApcFu5mawUooMhsH/1bYdRMn1PTlmRiPRJkoookkWOhVH\n" +
  "703iYVCkZL39DF9tXG4iANNdjFepVIcSZduyZBXslAvuvFWaNu9I52IYyDYDadDT\n" +
  "oLvXFi3dsfnIrGf68JLBqZnOSxVLg80Ht3maOSbc36DbUO5YE9nU0nhA/wQ983j0\n" +
  "FDzBKzrPYPFNIbmI3SEVPjhrFrxci7+Tbdky4SpIwLRN7SWpbu/GPmLzlvD8XMuW\n" +
  "cw1dsBoqV9w+pMwy2DTctQ3SU3nbxuvMgMKu4wuX9mWLiCmykGcLvYUrI8CKFzzP\n" +
  "f48Qd2/HC1lzwCKiwn57MhokbzcKWhB8E96X3HGtVwIDAQABo1MwUTAdBgNVHQ4E\n" +
  "FgQUPsTW/85jbxCCvyiMwx33xM6nh/IwHwYDVR0jBBgwFoAUPsTW/85jbxCCvyiM\n" +
  "wx33xM6nh/IwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEANfYP\n" +
  "er5WrTFuWtY2ApxMjvDWo0xuL6+kCv4UkaLtfsR4TkcF9svGtb/PvrlOmffP+By1\n" +
  "TkBoMju3uHU7/hPo5UMRUI445QnIVSSpLaHnLHBi85KGMl55ErRbOGVgYR1n+Qxo\n" +
  "nFwucn2hjHBmswXNY9DJx8vmewKqa9Kc694YxNx44tmMNbokUYKliofnisxxhM3Y\n" +
  "4REA+w3bme7z7JDLt0J1EGCAEeT2NcLbnXw8gEKJ0D39Zt5b5qR1Re4wWtknXlkc\n" +
  "5VGw96c5jbejD49BJsDWqVsJG1D4dhJ61ybBT6j2aRnSH+MxPHpH/Rm9wXIGha4G\n" +
  "mZLuodbHDBrL/FT4cQ==\n" +
  "-----END CERTIFICATE-----\n";
