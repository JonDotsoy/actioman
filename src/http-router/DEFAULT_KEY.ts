if (process.env.NODE_ENV !== "test") {
  console.warn("The variable DEFAULT_KEY is only to test environment");
}

export const DEFAULT_KEY =
  "-----BEGIN PRIVATE KEY-----\n" +
  "MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQClwW7mZrBSigyG\n" +
  "wf/Vth1EyfU9OWZGI9EmSiiiSRY6FUfvTeJhUKRkvf0MX21cbiIA012MV6lUhxJl\n" +
  "27JkFeyUC+68VZo270jnYhjINgNp0NOgu9cWLd2x+cisZ/rwksGpmc5LFUuDzQe3\n" +
  "eZo5JtzfoNtQ7lgT2dTSeED/BD3zePQUPMErOs9g8U0huYjdIRU+OGsWvFyLv5Nt\n" +
  "2TLhKkjAtE3tJalu78Y+YvOW8Pxcy5ZzDV2wGipX3D6kzDLYNNy1DdJTedvG68yA\n" +
  "wq7jC5f2ZYuIKbKQZwu9hSsjwIoXPM9/jxB3b8cLWXPAIqLCfnsyGiRvNwpaEHwT\n" +
  "3pfcca1XAgMBAAECggEAGE8voAvEZT6IysSt7B5a2YlS9aky5it/k4GBxxc38snH\n" +
  "NIH0JviIO5wffg6Kv+kJBJz7FcePdm2e1aBsX8BVY3sChs15IgBi7UhWQRqm0oCF\n" +
  "YQfbag3M+zqaYyqWw2yKiYCcoCdd3VUXAiIL68DW2NdDAvIJqdOjIkDvq1bobQn+\n" +
  "MNdNBnu58t95JRfEbQHiDUUVTGv1b4E8gAUtCmsDPtsSQAox8rTD4UTA0+9oxE5e\n" +
  "SoV69YWS+BqFaDZgxxTKxyoL6g4K7WrF58oOSIlayUBGOO0VpGBf3sFB9YvYw76P\n" +
  "7m8S1EY71G8/Bp4JoVrNR7TemG0Qotc/3AapEjyv7QKBgQDVKfOWe81oKuFJlpKP\n" +
  "1W0v7xoo0fV9GXJlOrpDTvcuoY+XiMuQtTWDl7de+4y+mzQ+YUWKCrEn6yj0eUgy\n" +
  "owEpk8p6E+nvU0w9Yhnlk0wmfHkQw/P4kwCuG7sFvcpJ8471Ad94NWLddeHdVQZb\n" +
  "3A8PZgXouoBstTlKJ9WkZHxmCwKBgQDHEJmiR/U6m882A8G7yhoDB4U2tYPS2WZq\n" +
  "sZdgaCDrYZNME46N11b6HEJYD6jHyP33FPgyLD9uz157/1WTod6y2fdrlRnTWXZt\n" +
  "KKX837tOaKZFYpJtTzVFdR6nCO6O4fL6tK/CyNIWOB+kAXmtr8GBfbBfGHuePuYr\n" +
  "9jjkGmIhZQKBgBes7Cgnof3S3Wjd93NQQitTxDUKdNirLeB+cFevepbMTPWItQti\n" +
  "hQoTnMNI7y3wcRIInb+iRmDnponYuqMdJbDngFKkyZPSTaJ3teagwJF5u7yX9w4B\n" +
  "zc6It/IATWp/Sb5+scJZM5zjTe3WVrl3LJkKPNfHdyIb0rwgxoLWwlZXAoGALI8s\n" +
  "tMTCwgHKml6w6SZOOEtgqGRtQ69iW5OxPOVSKR5IaIkgKeQAFjBpMWkf8MivALXy\n" +
  "8Ngyws0sazk2oLtlLEfaOHzBsSufdMPalAZI2T5XqcRs3ocSd5/F4sH/FIDDcRct\n" +
  "GL73N/4MJxwlbpmfCStnZYIHed6+iyysg3LccHkCgYBTPy0YejvOaTr3FMqIeL4r\n" +
  "3TsSbN9Q3/BX0oEvgTiJvLU1wsT3G0VzBE9qDbAmn9NuJhDKKgQRA7BdPVijZwib\n" +
  "NS49DtF2a5rB2wXfWNTG4KsQUPQ+zkhu3TyFUQijdlD+Vh47vD3bnH8hgInllUNc\n" +
  "WWPRkIP/TR8122xeYnv/hQ==\n" +
  "-----END PRIVATE KEY-----\n";
