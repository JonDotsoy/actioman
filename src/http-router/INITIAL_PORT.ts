export const INITIAL_PORT = 30320;
