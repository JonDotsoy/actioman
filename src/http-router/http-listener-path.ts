export const HTTP_LISTENER_PATH = new URL(
  "./http-listener.js",
  import.meta.url,
);
