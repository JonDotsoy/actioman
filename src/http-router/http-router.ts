import { Router } from "artur";
import { Actions, type Action } from "../actions/actions.js";
import { z } from "zod";
import { get } from "@jondotsoy/utils-js/get";
import { result } from "@jondotsoy/utils-js/result";
import { actionsToJson } from "../exporter-actions/exporter-actions.js";
import { Configs, type ConfigsModule } from "../configs/configs.js";
import { MetricsClient } from "../metric/metrics-client/metrics-client.js";

const isParser = (value: any): value is { parse: (value: any) => any } =>
  get.function(value, "parse") !== undefined;
const isZodType = (value: any): value is z.ZodType =>
  value instanceof z.ZodType;
const isSSEAction = (
  action: Action<any, any, any>,
): action is Action<any, any, true> => action.sse === true;

class EventStreamDataEncoder {
  encode(data: { id?: string; event?: string; data: any }) {
    const l = [];
    if (data.id) l.push(`id: ${data.id}`);
    if (data.event) l.push(`event: ${data.event}`);
    l.push(`data: ${JSON.stringify(data.data)}`);
    return new TextEncoder().encode(l.join("\n") + "\n\n");
  }
}

export class HTTPRouter {
  router: Router<"pass">;
  metrics: MetricsClient;

  constructor(
    actions: Actions,
    readonly configs?: Configs,
  ) {
    const actionsJson = actionsToJson(actions);

    this.metrics = new MetricsClient();

    this.router = new Router({
      errorHandling: "pass",
      middlewares: [
        ...(configs?.httpListenerMiddlewares ?? []),
        (fetch) => async (req) => {
          const res = await fetch(req);
          res?.headers.set("X-Powered-By", "Actioman");
          res?.headers.set("X-Actioman-Version", "v1.0.0");
          return res;
        },
      ],
    });

    configs?.httpSetup(this);

    this.router.use("GET", `/__actions`, {
      fetch: () => Response.json({ actions: actionsJson }),
    });

    for (const [name, describe] of Object.entries(Actions.describe(actions))) {
      const metricLabels = {
        action: name,
      };

      const requestErrorCountersMetric = this.metrics.counter({
        name: "request_error_counters",
        labels: metricLabels,
      });
      const requestCountersMetric = this.metrics.counter({
        name: "request_counters",
        labels: metricLabels,
      });
      const requestDataTransferenceBytesMetric = this.metrics.counter({
        name: "request_data_transference_bytes",
        labels: metricLabels,
      });
      const requestDurationSecondsMetric = this.metrics.averageBySecond({
        name: "request_duration_seconds",
        labels: metricLabels,
      });

      const isSSE = isSSEAction(describe);
      const method = "POST";
      this.router.use(method, `/__actions/${name}`, {
        middlewares: [
          (fetch) => (req) => {
            const start = Date.now();
            requestCountersMetric.increment();
            return Promise.resolve(fetch(req))
              .catch((err) => {
                requestErrorCountersMetric.increment();
                throw err;
              })
              .finally(() => {
                requestDurationSecondsMetric.add(Date.now() - start);
              });
          },
        ],
        fetch: async (req) => {
          if (isSSE) return this.sseHandler(describe, req);
          return this.handler(describe, req);
        },
      });
    }
  }

  private async sseHandler(action: Action<any, any, true>, req: Request) {
    const url = new URL(req.url);
    const inputSearchParams = url.searchParams.get("input") ?? null;
    const [, data] = await result(() =>
      inputSearchParams ? JSON.parse(inputSearchParams) : null,
    );
    const input = isParser(action.input) ? action.input.parse(data) : null;
    const resultAction = await action.handler(input);
    const body = new ReadableStream<Uint8Array>({
      pull: async (ctrl) => {
        const [err, iteratorResult] = await result(() => resultAction.next());

        if (err) {
          console.error(err);
          ctrl.enqueue(
            new EventStreamDataEncoder().encode({
              event: "error",
              data: "Internal error",
            }),
          );
          return ctrl.close();
        }

        const { done, value } = iteratorResult;

        if (done) {
          ctrl.enqueue(
            new EventStreamDataEncoder().encode({
              event: "close",
              data: true,
            }),
          );
          return ctrl.close();
        }

        const output = isParser(action.output)
          ? action.output.parse(value)
          : value;

        ctrl.enqueue(new EventStreamDataEncoder().encode({ data: output }));
      },
    });
    return new Response(body, {
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
        "X-Accel-Buffering": "no",
      },
    });
  }

  private async handler(action: Action<any, any, any>, req: Request) {
    const [, data] = await result(() => req.json());
    const input = isParser(action.input) ? action.input.parse(data) : null;
    const resultDescribe = await action.handler(input);
    const output = isParser(action.output)
      ? action.output.parse(resultDescribe)
      : resultDescribe;
    return Response.json(output);
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const actions = Actions.fromModule(module, configs);
    if (!actions) throw new Error(`No actions found in ${module}`);
    return new HTTPRouter(actions, Configs.fromModule(configs));
  }
}
