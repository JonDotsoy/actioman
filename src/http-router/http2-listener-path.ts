export const HTTP2_LISTENER_PATH = new URL(
  "./http2-listener.js",
  import.meta.url,
);
