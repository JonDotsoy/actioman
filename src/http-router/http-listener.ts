import * as http from "http";
import { HTTPRouter } from "./http-router.js";
import net from "net";
import type { ConfigsModule } from "../configs/configs.js";

type ListenOptions = {
  silent: boolean;
};

const sanatizeHostname = (hostname: string) => {
  switch (hostname) {
    case "::1":
    case "::":
    case "0.0.0.0":
    case "127.0.0.1":
    case "::ffff:127.0.0.1":
      return "localhost";
  }

  return hostname;
};

const INITIAL_PORT = 30_320;

const findNextPort = async () => {
  let porposalPort = INITIAL_PORT;
  while (true) {
    porposalPort++;
    if (porposalPort >= INITIAL_PORT + 10_000)
      throw new Error("No available port");
    const port = await new Promise<number | null>((resolve) => {
      const connectiong = net.connect({
        host: "localhost",
        port: porposalPort,
      });

      connectiong.addListener("connect", () => {
        resolve(null);
        connectiong.destroy();
      });
      connectiong.addListener("error", (err) => {
        if ("code" in err && err.code === "ECONNREFUSED") resolve(porposalPort);
      });
    });
    if (typeof port === "number") return port;
  }
};

export class HTTPLister {
  server: http.Server;

  constructor(readonly httpRouter: HTTPRouter) {
    this.server = http.createServer(async (req, res) => {
      try {
        if (!(await httpRouter.router.requestListener(req, res))) {
          res.statusCode = 404;
          return res.end();
        }
      } catch (error) {
        console.error(error);
        res.statusCode = 500;
        return res.end();
      }
    });
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      if (this.server.address() === null) return resolve(null);
      this.server.close((err) => {
        if (err) return reject(err);
        return resolve(null);
      });
    });
  }

  async listen(port?: number, hostname?: string, options?: ListenOptions) {
    const silent = options?.silent ?? HTTPLister.defaultOptions.silent;

    const log = (message: string) => {
      if (silent) return;
      console.log(message);
    };

    const portToListen = port ?? (await findNextPort());
    const hostnameToListen = hostname ?? "localhost";

    const url = await new Promise<URL>((resolve) => {
      this.server.listen(portToListen, hostnameToListen, () => {
        const address = this.server.address();
        if (typeof address === "object" && address !== null)
          return resolve(
            new URL(
              `http://${sanatizeHostname(address.address)}:${address.port}`,
            ),
          );
        return resolve(
          new URL(
            `http://${sanatizeHostname(hostnameToListen)}:${portToListen}`,
          ),
        );
      });
    });

    if (!silent) {
      for (const route of this.httpRouter.router.routes) {
        const method = route.method;
        const urlPattern = route.urlPattern;
        log(`Route ${method} ${urlPattern.pathname}`);
      }
    }
    log(`Listening on ${url.toString()}`);

    return url;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const httpRouter = HTTPRouter.fromModule(module, configs);
    if (!httpRouter) throw new Error("No actions found");
    return new HTTPLister(httpRouter);
  }

  static defaultOptions = {
    silent: false,
  };
}
