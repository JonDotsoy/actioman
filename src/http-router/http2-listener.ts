import { HTTPRouter } from "./http-router.js";
import type { ConfigsModule } from "../configs/configs.js";
import { sanitizeHostname } from "./utils/sanitize-hostname.js";
import { findNextPort } from "./utils/find-next-port.js";
import * as http2 from "http2";
import { DEFAULT_CERT } from "./DEFAULT_CERT.js";
import { DEFAULT_KEY } from "./DEFAULT_KEY.js";
import { requestHttp2ToRequest } from "./utils/request-http2-to-request.js";

type ListenOptions = {
  silent: boolean;
};

export const INITIAL_PORT = 30_320;

export class HTTP2Lister {
  server: http2.Http2Server;
  url: URL | undefined;

  constructor(readonly httpRouter: HTTPRouter) {
    this.server = http2.createSecureServer(
      {
        key: DEFAULT_KEY,
        cert: DEFAULT_CERT,
        settings: { initialWindowSize: 8 * 1024 * 1024 },
      },
      async (requestHttp2, responseHttp2) => {
        try {
          // console.log(r)

          const response = await this.httpRouter.router.fetch(
            requestHttp2ToRequest(requestHttp2),
          );

          if (!response) {
            responseHttp2.statusCode = 404;
            return responseHttp2.end();
          }

          responseHttp2.statusCode = response.status;
          response.headers.forEach((value, key) => {
            if (key.startsWith(":")) return;
            if (responseHttp2.hasHeader(key)) {
              responseHttp2.appendHeader(key, value);
            } else {
              responseHttp2.setHeader(key, value);
            }
          });
          if (response.body) {
            const reader = response.body.getReader();
            for (
              let readResult = await reader.read();
              !readResult.done && !requestHttp2.aborted;
              readResult = await reader.read()
            ) {
              responseHttp2.write(readResult.value);
            }
          }
          return responseHttp2.end();
        } catch (error) {
          console.error(error);
          responseHttp2.statusCode = 500;
          return responseHttp2.end();
        }
      },
    );
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      if (this.server.address() === null) return resolve(null);
      this.server.close();
      return resolve(null);
    });
  }

  async listen(port?: number, hostname?: string, options?: ListenOptions) {
    const silent = options?.silent ?? HTTP2Lister.defaultOptions.silent;

    const log = (message: string) => {
      if (silent) return;
      console.log(message);
    };

    const portToListen = port ?? (await findNextPort());
    const hostnameToListen = hostname ?? "::";

    const url = await new Promise<URL>((resolve) => {
      this.server.listen(portToListen, hostnameToListen, () => {
        const address = this.server.address();
        if (typeof address === "object" && address !== null)
          return resolve(
            new URL(
              `https://${sanitizeHostname(address.address)}:${address.port}`,
            ),
          );
        return resolve(
          new URL(
            `https://${sanitizeHostname(hostnameToListen)}:${portToListen}`,
          ),
        );
      });
    });

    if (!silent) {
      for (const route of this.httpRouter.router.routes) {
        const method = route.method;
        const urlPattern = route.urlPattern;
        log(`Route ${method} ${urlPattern.pathname}`);
      }
    }
    log(`Listening on ${url.toString()}`);

    this.url = url;

    return url;
  }

  static fromModule(module: unknown, configs?: ConfigsModule) {
    const httpRouter = HTTPRouter.fromModule(module, configs);
    if (!httpRouter) throw new Error("No actions found");
    return new HTTP2Lister(httpRouter);
  }

  static defaultOptions = {
    silent: false,
  };
}

export const HTTPLister = HTTP2Lister;
