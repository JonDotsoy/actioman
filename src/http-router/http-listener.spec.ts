import { afterAll, afterEach, beforeAll, describe, expect, it } from "bun:test";
import { HTTPLister } from "./http-listener";
import { HTTPRouter } from "./http-router";
import { CleanupTasks } from "@jondotsoy/utils-js/cleanuptasks";
import { DEFAULT_CERT } from "./DEFAULT_CERT";
import { DEFAULT_KEY } from "./DEFAULT_KEY";
import * as https from "https";
import * as http from "http";
import { EventSource } from "eventsource";

let port = 7080;

describe("HTTPLister", () => {
  it("should return 404 for unknown paths", async () => {
    await using cleanupTasks = new CleanupTasks();
    const httpLister = HTTPLister.fromModule({});
    cleanupTasks.add(() => httpLister.close());
    const url = await httpLister.listen(port++);

    const res = await fetch(new URL("/", url!));

    expect(res.status).toEqual(404);
  });

  it("should return 200 for /__actions", async () => {
    await using cleanupTasks = new CleanupTasks();
    const httpLister = HTTPLister.fromModule({});
    cleanupTasks.add(() => httpLister.close());
    const url = await httpLister.listen(port++);

    const res = await fetch(new URL("/__actions", url!));

    expect(res.status).toEqual(200);
  });

  it("should start a server", async () => {
    await using cleanupTasks = new CleanupTasks();
    const { default: express } = await import("express");
    const app = express();

    const httpRouter = HTTPRouter.fromModule({ hi: () => "ok" });

    app.use("/", async (req, res, next) => {
      const ok = await httpRouter.router.requestListener(req, res);
      if (!ok) next();
    });

    const server = app.listen(port++);
    cleanupTasks.add(
      () => new Promise<true>((resolve) => server.close(() => resolve(true))),
    );
  });

  describe("integration tests", () => {
    let url = new URL(`http://localhost:${port++}`);

    const cleanupTasks = new CleanupTasks();
    beforeAll(async () => {
      const { default: express } = await import("express");
      const app = express();

      const httpRouter = HTTPRouter.fromModule({ hi: () => "ok" });

      app.use("/", async (req, res, next) => {
        const ok = await httpRouter.router.requestListener(req, res);
        if (!ok) next();
      });

      const server = app.listen(url.port);
      cleanupTasks.add(
        () => new Promise<true>((resolve) => server.close(() => resolve(true))),
      );
    });

    afterAll(async () => {
      await cleanupTasks.cleanup();
    });

    it("should return 200 for /__actions", async () => {
      const res = await fetch(new URL("/__actions", url));
      expect(res.status).toEqual(200);
      expect(await res.text()).toMatchSnapshot();
    });

    it("should return 200 for /__actions/hi", async () => {
      const res = await fetch(new URL("/__actions/hi", url), {
        method: "POST",
      });
      expect(res.status).toEqual(200);
      expect(await res.text()).toMatchSnapshot();
    });
  });

  it("should create a server with actions", async () => {
    await using cleanupTasks = new CleanupTasks();
    const httpLister = HTTPLister.fromModule({
      biz: () => true,
      foo: () => true,
      taz: () => true,
    });
    cleanupTasks.add(() => httpLister.close());
    await httpLister.listen(port++);
  });

  it("should create a server with ssl", async () => {
    await using cleanupTasks = new CleanupTasks();
    const httpLister = HTTPLister.fromModule(
      {
        hi: () => "",
      },
      {
        server: {
          ssl: {
            key: DEFAULT_KEY,
            cert: DEFAULT_CERT,
          },
        },
      },
    );
    cleanupTasks.add(() => httpLister.close());
    const url = await httpLister.listen(port++);

    expect(url.protocol).toEqual("https:");

    const res = await new Promise<http.IncomingMessage>((resolve, reject) => {
      process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
      https
        .request(new URL("./__actions", url))
        .end()
        .addListener("connect", () => {
          console.log("connect");
        })
        .addListener("close", () => {
          console.log("end");
        })
        .addListener("error", (err) => {
          reject(err);
        })
        .addListener("response", (res) => {
          resolve(res);
        });
    });

    expect(res.statusCode).toEqual(200);
  });

  it("should handle sse actions", async () => {
    await using cleanupTasks = new CleanupTasks();
    const httpLister = HTTPLister.fromModule({
      *hi() {
        yield 1;
        yield 2;
      },
    });
    cleanupTasks.add(() => httpLister.close());

    const url = await httpLister.listen(port++);

    const eventSource = new EventSource(new URL("/__actions/hi", url), {
      fetch: (url, init) => fetch(url, { ...init, method: "POST" }),
    });

    const messages: any[] = [];

    eventSource.addEventListener("message", (message) => {
      messages.push(message.data);
    });

    await new Promise<void>((resolve, reject) => {
      eventSource.addEventListener("error", reject);
      eventSource.addEventListener("close", () => {
        resolve();
      });
    });

    expect(messages).toEqual([`1`, `2`]);
  });
});
