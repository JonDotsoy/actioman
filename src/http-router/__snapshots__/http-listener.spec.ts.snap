// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`HTTPLister integration tests should return 200 for /__actions 1`] = `"{"actions":{"hi":{"description":null,"sse":false,"input":null,"output":null}}}"`;

exports[`HTTPLister integration tests should return 200 for /__actions/hi 1`] = `""ok""`;
